external/fv3fit/fv3fit/reservoir/train.py

```diff
@@ -0,0 +1,213 @@
+    stack_time_series_samples,
+    concat_variables_along_feature_dim,
+)
+
+# allow reshaping of tensor data
+from tensorflow.python.ops.numpy_ops import np_config
+
+np_config.enable_numpy_behavior()
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+def _add_input_noise(arr: np.ndarray, stddev: float) -> np.ndarray:
+    return arr + np.random.normal(loc=0, scale=stddev, size=arr.shape)
+
+
+def _encode_columns(data: tf.Tensor, encoder: tf.keras.Model,) -> np.ndarray:
+    # reduce N x M x V dim data to N x M x Z dim
+    # where V is original number of features (usually
+    # variables * vertical levels) and Z << V is a smaller
+    # number of latent dimensions
+    original_sample_shape = data.shape[:-1]
+    original_z_dim = data.shape[-1]
+
+    reshaped = data.reshape(-1, original_z_dim)
+    encoded_reshaped = encoder.predict(reshaped)
+    return encoded_reshaped.reshape(*original_sample_shape, -1)
+
+
+@register_training_function("pure-reservoir", ReservoirTrainingConfig)
+def train_reservoir_model(
+    hyperparameters: ReservoirTrainingConfig,
+    train_batches: tf.data.Dataset,
+    validation_batches: Optional[tf.data.Dataset],
+) -> Predictor:
+
+    if hyperparameters.autoencoder_path is not None:
+        autoencoder = Autoencoder.load(hyperparameters.autoencoder_path)
+    else:
+        autoencoder = None  # type: ignore
+
+    # Standard scale data, calculating mean and std along the stacked x,y,z,feature dim
+    norm_batch = next(iter(train_batches))
+    norm_batch_concat = concat_variables_along_feature_dim(
+        variables=hyperparameters.input_variables, variable_tensors=norm_batch
+    )
+
+    sample_batch = (
+        _encode_columns(norm_batch_concat, autoencoder.encoder)
+        if autoencoder is not None
+        else norm_batch_concat
+    )
+    subdomain_config = hyperparameters.subdomain
+    rank_divider = RankDivider(
+        subdomain_layout=subdomain_config.layout,
+        rank_dims=subdomain_config.rank_dims,
+        rank_extent=tuple(sample_batch.shape),
+        overlap=subdomain_config.overlap,
+    )
+
+    # first data dim is time, the rest of the elements of the
+    # subdomain+halo are are flattened into feature dimension
+    n_input_features = int(
+        np.prod(rank_divider.get_subdomain_extent(with_overlap=True)[1:])
+    )
+    reservoir = Reservoir(
+        hyperparameters=hyperparameters.reservoir_hyperparameters,
+        input_size=n_input_features,
+    )
+
+    # One readout is trained per subdomain when iterating over batches,
+    # and they are combined after training into a CombinedReadout
+    subdomain_regressors = [
+        BatchLinearRegressor(hyperparameters.readout_hyperparameters)
+        for r in range(rank_divider.n_subdomains)
+    ]
+    for b, batch_data in enumerate(train_batches):
+        # reservoir increment occurs in this call, so always call this
+        # function even if X, Y are not used for readout training.
```

Comment:
I guess this is referring to the call to `_get_reservoir_state_time_series` call.  Maybe move the comment farther down?